### PR TITLE
Autonomous/Drive Control State Machine

### PR DESCRIPTION
**Pull request by StephenWelch**
Allows the driver to take control of the robot during autonomous and run commands during teleop while a button is held. Also includes fixes required to make unit testing work.

### CODE REVIEW

**Review by DanielTolbert — APPROVED** [review]

**Review by DanielTolbert — DISMISSED** [review]

File: robot/src/main/java/us/ilite/lib/drivers/Clock.java [hunk 1]
```diff
@@ -14,17 +14,22 @@
 
     private ILog mLogger = Logger.createLog(Clock.class);
 
+    private double mStartTime = 0.0;
     private double mCurrentTime = 0.0;
     private boolean hasTimeUpdatedThisCycle = false;
     private boolean mIsSimulated = false;
 
+    public Clock() {
```
Comment (oconnelc) [inline robot/src/main/java/us/ilite/lib/drivers/Clock.java:22]: Don't need this constructor

File: robot/src/main/java/us/ilite/lib/drivers/Clock.java [hunk 2]
```diff
@@ -41,15 +46,25 @@ public void cycleEnded() {
...
         }
     }
 
     public Clock simulated() {
+        mStartTime = getJavaTime();
         mIsSimulated = true;
         return this;
     }
 
+    private double getJavaTime() {
```
Comment (oconnelc) [inline robot/src/main/java/us/ilite/lib/drivers/Clock.java]: Recommend making this a static method

File: robot/src/main/java/us/ilite/lib/drivers/Clock.java [hunk 3]
```diff
@@ -41,15 +46,25 @@ public void cycleEnded() {
...
+        return System.currentTimeMillis() / 1000.0;
+    }
+
+    private double getRobotTime() {
```
Comment (oconnelc) [inline robot/src/main/java/us/ilite/lib/drivers/Clock.java]: Recommend making this a static method

File: robot/src/main/java/us/ilite/robot/Robot.java [hunk 4]
```diff
@@ -91,26 +91,24 @@ public void autonomousInit() {
         initTimer.start();
         mLogger.info("Starting Autonomous Initialization...");
 
+        mSuperstructure.startCommands(new FollowTrajectory(trajectory, mDrive, true));
+//        mCommandQueue.startCommands(new CharacterizeDrive(mDrive, false, false));
```
Comment (oconnelc) [inline robot/src/main/java/us/ilite/robot/Robot.java:95]: Can you take this line out?

File: robot/src/main/java/us/ilite/robot/commands/Delay.java [hunk 5]
```diff
@@ -1,10 +1,14 @@
 package us.ilite.robot.commands;
 
+import com.flybotix.hfr.util.log.ILog;
+import com.flybotix.hfr.util.log.Logger;
+
 public class Delay implements ICommand {
+	private ILog mLog = Logger.createLog(Delay.class);
```
Comment (oconnelc) [inline robot/src/main/java/us/ilite/robot/commands/Delay.java:7]: Recommend making this static

File: robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java [hunk 6]
```diff
@@ -11,46 +13,51 @@
...
 import us.ilite.robot.modules.Module;
-
-import java.util.LinkedList;
-import java.util.Queue;
+import us.ilite.robot.modules.Superstructure;
 
 public class DriverInput extends Module {
 
     private static final double DRIVER_SUB_WARP_AXIS_THRESHOLD = 0.5;
+    private ILog mLog = Logger.createLog(DriverInput.class);
```
Comment (oconnelc) [inline robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:25]: Recommend making this static

File: robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java [hunk 7]
```diff
@@ -11,46 +13,51 @@
...
         this.driveTrain = pDrivetrain;
+        this.mSuperstructure = pSuperstructure;
         this.mData = pData;
-        this.desiredCommandQueue = new LinkedList<>();
         this.mDriverInputCodex = mData.driverinput;
         this.mOperatorInputCodex = mData.operatorinput;
-        this.mDriverJoystick = new Joystick(0);
-        this.mOperatorJoystick = new Joystick(1);
+        if(pSimulated) {
+            // Use a different joystick library?
```
Comment (oconnelc) [inline robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:46]: Are things going to break if the mDriverJoystick and mOperatorJoystick are null? Maybe use Optional?

File: robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java [hunk 8]
```diff
@@ -61,87 +68,108 @@ public void periodicInput(double pNow) {
...
+           mData.driverinput.isSet(DriveTeamInputMap.DRIVER_SUB_WARP_AXIS)) {
+            double rotate = mData.driverinput.get(DriveTeamInputMap.DRIVER_TURN_AXIS);
+            double throttle = -mData.driverinput.get(DriveTeamInputMap.DRIVER_THROTTLE_AXIS);
+
+//		    throttle = EInputScale.EXPONENTIAL.map(throttle, 2);
+            rotate = EInputScale.EXPONENTIAL.map(rotate, 2);
+            rotate = Util.limit(rotate, 0.7);
+
+            if (mData.driverinput.get(DriveTeamInputMap.DRIVER_SUB_WARP_AXIS) > DRIVER_SUB_WARP_AXIS_THRESHOLD) {
+                throttle *= SystemSettings.kSnailModePercentThrottleReduction;
```
Comment (oconnelc) [inline robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:111]: Will throttle ever get too big or too small with this logic?

File: robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java [hunk 9]
```diff
@@ -61,87 +68,108 @@ public void periodicInput(double pNow) {
...
-            rotate *= SystemSettings.SNAIL_MODE_ROTATE_LIMITER;
+    /**
+     * Commands the superstructure to start vision tracking depending on
+     * button presses.
+     */
+    protected void updateVisionCommands() {
+
+        ETrackingType trackingType = null;
+        // Switch the limelight to a pipeline and track
+        if(mDriverInputCodex.isSet(DriveTeamInputMap.DRIVER_TRACK_TARGET_BTN)) {
```
Comment (oconnelc) [inline robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:131]: Is it possible to have multiple targets selected? If so we need to handle this better.

File: robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java [hunk 10]
```diff
@@ -61,87 +68,108 @@ public void periodicInput(double pNow) {
...
-//			System.out.println("LEFT: " + desiredLeftOutput +"\tRIGHT: " +  desiredRightOutput + "");
-//		}
+    }
 
-        driveTrain.setDriveMessage(new DriveMessage(desiredLeftOutput, desiredRightOutput, ControlMode.PercentOutput).setNeutralMode(NeutralMode.Brake));
+    public boolean isDriverAllowingTeleopCommands() {
+        boolean runCommands = false;
+        for(ELogitech310 l : SystemSettings.kTeleopCommandTriggers) {
+            if(mDriverInputCodex.isSet(l)) {
+                runCommands = true;
```
Comment (oconnelc) [inline robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:159]: Recommend breaking

File: robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java [hunk 11]
```diff
@@ -61,87 +68,108 @@ public void periodicInput(double pNow) {
...
+            }
+        }
+        return runCommands;
+    }
 
+    public boolean isAutoOverridePressed() {
+        boolean autonOverride = false;
+        for(ELogitech310 l : SystemSettings.kAutonOverrideTriggers) {
+            if(mDriverInputCodex.isSet(l) && mDriverInputCodex.get(l) > SystemSettings.kAutonOverrideAxisThreshold) {
+                autonOverride = true;
```
Comment (oconnelc) [inline robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:169]: Recommend breaking

File: robot/src/test/java/us/ilite/robot/driverinput/DriverInputTest.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+import us.ilite.robot.modules.Superstructure;
+
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.*;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DriverInputTest {
+
+    // We're only testing integration between Superstructure and DriverInput, so we can mock this
+    @Mock private Drive mDrive;
```
Comment (oconnelc) [inline robot/src/test/java/us/ilite/robot/driverinput/DriverInputTest.java:26]: One problem with the mock annotations is that once the object is created, it's never re-created on subsequent tests. You'll want to have an "@after" annotation to null out the member,.

**Review by oconnelc — CHANGES_REQUESTED** [review]
I have a couple of comments that should be looked at. Nothing too critical, but just some good housekeeping.

File: robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java (hunk 7 above)
Comment (flybotix) [reply robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:46]: They aren't null; they are codexes instantiated at robot init.

File: robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java (hunk 8 above)
Comment (flybotix) [reply robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:111]: Technically, throttle can be oversaturated.  Oversaturation is handled in the SRX if we don't handle it ourselves.

The throttle reduction constant reduces power.  Oversaturation isn't an issue here.

File: robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java (hunk 9 above)
Comment (flybotix) [reply robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:131]: I created a Trello card to handle this scenario.

**Review by flybotix — APPROVED** [review]

**Review by oconnelc — APPROVED** [review]

**Pull request merged**